docs/concept/#894-Efficient-orchestration-of-edc-negotiations/#894-Efficient-orchestration-of-edc-negotiations.md

```diff
@@ -0,0 +1,103 @@
+# #894 Efficient orchestration of edc negotiations
+
+| Key           | Value                                                                    |
+|---------------|--------------------------------------------------------------------------|
+| Creation date | 09.10.2024                                                               |
+| Ticket Id     | https://github.com/eclipse-tractusx/item-relationship-service/issues/894 |    
+| State         | DONE                                                                     | 
+
+## Table of Contents
+
+1. [Decision](#Decision)
+2. [Rationale](#Rationale)
+3. [Approach](#Approach)
+4. [Testing Strategy](#Testing-Strategy)
+
+## Decision
+
+Implementation of a new EDC Orchestrator within the EDC Client Library for efficient management of negotiations. The
+goal is to prevent duplicate negotiations for assets, thereby ensuring more efficient processing in asset consumption.
+
+## Rationale
+
+Each job is executed in isolation. This results in many jobs conducting a multitude of unnecessary contract negotiations
+for the same DT Registry Asset in the EDC catalog. In case the same catalog offer is requested in different jobs this
+leads to not performant overall executions. Therefore, a concept is required how to handle this in the most efficient
+way.
+
+## Approach
+
+A new service “NegotiationOrchestrator” is introduced which is a middle layer between the already existing components
+EdcSubmodelClientImpl and ContractNegotiationService, EDCCatalogFacade and EndpointDataReferenceCacheService.
+
+The service handles the following tasks:
+
+- orchestration of catalog requests by id
+- orchestration of catalog requests by type
+- orchestration of contract negotiations
+- caching of ongoing contract negotiations
+- caching of EDR tokens
+
+To prevent too many parallel requests to the EDC, an ExecutorService handles the catalog requests and negotiations with
+a defined thread pool.
+
+To cache the currently ongoing negotiations, a ConcurrentHashMap with EDC AssetIds as key can be used. Due to the
+asynchronous process within NegotiationOrchestrator, the future EDR can be created and added to the cache as soon as the
+method is called and will be removed only after the negotiation is completed. In addition, a time to live can be added
+for cleanup.
+
+To cache EDR tokens, the EndpointDataReferenceCacheService can be used. The caching service wraps EDR tokens into an
```

Comment:
Have you though about compression of tokens?

Reply:
This is a highly critical path. How can we ensure that this mechanism functions comprehensively and without errors? I see a potential issue where tokens that are no longer valid might still be treated as valid due to failure scenarios within the system. This poses a significant risk to the acceptance of the system, as security mechanisms could be compromised.

Additionally, is it possible to externally manipulate this mechanism to assign tokens a longer validity period than what is actually stored in the EDC? If so, this could introduce severe security vulnerabilities. Please clarify how these risks are mitigated and ensure that the mechanism is robust against such scenarios.

Reply:
The validation of the Token is still done by the EDC. If IRS could somehow change the validity period, that would be a security issue on EDC side. 
The validity period assigned by IRS is just the tokens own expiry date to simplify token handling during IRS processing. If a Token is wrongly labeled and used for a data exchange, even though it is expired, an authentication error is thrown by the EDC and IRS will retry the entire process, this time with a new token. 
All of this is already implemented as part of EndpointDataReferenceCacheService and not part of this concept.